### PR TITLE
Implement grouped convolutions

### LINKED ISSUE #114: `groups` support for Conv1D?
Hi, this is a really awesome project!

I'm trying to port a [model](https://arxiv.org/abs/2102.06200) that makes use of 1D convolutions but the immediate thing I ran into was that the Conv1D layer didn't have a parameter for `groups`, as present in PyTorch/Tensorflow. Learning resources on low-level NN programming is a little terse, but I'd like to tackle this!

My (high-level) understanding of it goes something along the lines of:

```py
from torchinfo import summary

x = torch.nn.Conv1d(9, 3, kernel_size=1, groups=1, bias=False)
y = torch.nn.Conv1d(9, 3, kernel_size=1, groups=3, bias=False)

summary(x)
summary(y)
```
Which gives:
```py
=================================================================
Layer (type:depth-idx)                   Param #
=================================================================
Conv1d                                   27
=================================================================
Total params: 27
Trainable params: 27
Non-trainable params: 0
=================================================================
=================================================================
Layer (type:depth-idx)                   Param #
=================================================================
Conv1d                                   9
=================================================================
Total params: 9
Trainable params: 9
Non-trainable params: 0
=================================================================
```
In the case of `x`, with `groups=1`, each input channel has its own 1x1 kernel. This is reflected in the weights:
```py
tensor([[[-0.2372],
         [-0.0259],
         [ 0.0075],
         [ 0.0602],
         [ 0.0205],
         [-0.2290],
         [-0.3304],
         [ 0.1302],
         [ 0.0697]],

        [[ 0.3148],
         [-0.1045],
         [ 0.2087],
         [ 0.2184],
         [ 0.2869],
         [-0.1255],
         [-0.0349],
         [ 0.2754],
         [ 0.0341]],

        [[-0.0065],
         [ 0.0904],
         [ 0.1445],
         [ 0.0337],
         [-0.0661],
         [ 0.2763],
         [-0.1375],
         [ 0.0841],
         [-0.0864]]]) torch.Size([3, 9, 1])
```
For each output channel, there are 9 1x1 kernels which correspond to each input channel.

Meanwhile, for the case of `y` with `groups=3`, it has the following weights:
```py
tensor([[[-0.0507],
         [ 0.1446],
         [ 0.1827]],

        [[-0.1260],
         [ 0.2465],
         [ 0.5095]],

        [[ 0.4771],
         [ 0.1377],
         [-0.0265]]]) torch.Size([3, 3, 1])
```
For each output channel, there are 3 1x1 kernels which correspond to each input channel **group**.

An intuitive way I've found to see how this works is:
```py
x.weight.requires_grad = False
y.weight.requires_grad = False

torch.nn.init.constant_(x.weight[0], 1.0)
torch.nn.init.constant_(x.weight[1], 0.9)
torch.nn.init.constant_(x.weight[2], 0.8)

torch.nn.init.constant_(y.weight[0], 1.0)
torch.nn.init.constant_(y.weight[1], 0.9)
torch.nn.init.constant_(y.weight[2], 0.8)

i = torch.ones(9, 2)
print(i)
print(x(i))
print(y(i))
```

and this yields:
```py
tensor([[1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.],
        [1., 1.]])
tensor([[9.0000, 9.0000],
        [8.1000, 8.1000],
        [7.2000, 7.2000]])
tensor([[3.0000, 3.0000],
        [2.7000, 2.7000],
        [2.4000, 2.4000]])
```

The result of `y(i)` yields significantly less "energy" than `x(i)` as each output channel now has less kernels to work with, 3 instead of 9.

### PR DESCRIPTION
**Pull request by purefunctor**
Closes #114. This adds the `groups_of` parameter and an alternative implementation for `forward` that works on grouped convolutions. I had to conditionally define `forward` depending on the value of `groups_of` since I had trouble reconciling the implementation with `groups_of=1`

### CODE REVIEW

**Comment by purefunctor** [comment]
TODO:
- [ ] Implement for non-templated Conv1D
- [ ] Implement for other backends

**Comment by jatinchowdhury18** [comment]
Nice! I haven't had time yet to go through the PR in detail, but at first glance, everything looks great! I wouldn't worry about the other backends just yet, and I'm happy to help with those implementation as well.

File: RTNeural/conv1d/conv1d.h [hunk 1]
```diff
@@ -190,10 +195,47 @@ class Conv1DT
...
+            }
+
+            for(int k = 0; k < kernel_size; ++k)
+            {
+                outs[i] = std::inner_product(
+                    weights[i][k].begin(),
+                    weights[i][k].end(),
+                    state_cols[k].begin(),
+                    outs[i]);
+            }
```
Comment (jatinchowdhury18) [inline RTNeural/conv1d/conv1d.h:229]: This is cool! Do you think it would make sense to merge these two loops, or maybe there's some reason the logic should be kept separate? I don't think it would make any difference from a performance perspective, so I'm more just thinking about what would make the most sense to a potential reader.

File: python/conv1d_torch_group.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import json
+import numpy as np
+import torch
+from torch import nn
+
+np.random.seed(1001)
+torch.manual_seed(0)
+
+f = nn.Conv1d(in_channels=6, out_channels=3, kernel_size=1, groups=3, dtype=torch.float64)
```
Comment (jatinchowdhury18) [inline python/conv1d_torch_group.py:9]: Nice! I was wondering if it might make sense to add a couple more layers, maybe with some different dilations, kernel sizes, etc, just to see if there's any edge cases that we'd need to handle differently.

File: tests/torch_conv1d_group_test.hpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        std::cout << "TESTING TORCH/CONV1D GROUP MODEL WITH DATA TYPE: DOUBLE" << std::endl;
+
+    const auto model_file = std::string { RTNEURAL_ROOT_DIR } + "models/conv1d_torch_group.json";
+    std::ifstream jsonStream(model_file, std::ifstream::binary);
+
+    nlohmann::json modelJson;
+    jsonStream >> modelJson;
+
+    RTNeural::ModelT<T, 6, 3, RTNeural::Conv1DT<T, 6, 3, 1, 1, 3, false>> model;
+    RTNeural::torch_helpers::loadConv1D<T>(modelJson, "", model.template get<0>());
```
Comment (jatinchowdhury18) [inline tests/torch_conv1d_group_test.hpp:53]: Do you know if TensorFlow also has a "groups" concept for their Conv1D layer? That might be a good thing to test for also, although we can add that in a future PR if it turns out to be complicated.

**Review by jatinchowdhury18 — COMMENTED** [review]
Looking good! I did a little bit of work on [another branch](https://github.com/jatinchowdhury18/RTNeural/tree/purefunctor-pure/implement-groups) mostly just rebasing from `main` and messing with the test code a little bit. I think this weekend I can go through and work more on the other implementations.

File: RTNeural/conv1d/conv1d.h (hunk 1 above)
Comment (purefunctor) [reply RTNeural/conv1d/conv1d.h:229]: I separated them initially to make it close to `groups_of=1` (and also for debugging purposes). I think merging them is fine now.

File: tests/torch_conv1d_group_test.hpp (hunk 3 above)
Comment (purefunctor) [reply tests/torch_conv1d_group_test.hpp:53]: I believe so, and it's functionally the same as PyTorch's

Comment (jatinchowdhury18) [reply tests/torch_conv1d_group_test.hpp:53]: Okay cool! I can take a look at adding a test case on the Tensorflow side.

**Comment by purefunctor** [comment]
I also pushed my attempt at implementing microTCN [9d6545c](https://github.com/jatinchowdhury18/RTNeural/pull/116/commits/9d6545c72368470ee3d52c8d195dee3a588e9990) but inference isn't correct yet. Do you reckon there's something I'm missing?

**Comment by jatinchowdhury18** [comment]
Nice work on getting the TCN working!

I think I have the necessary changes for the non-templated STL implementation and other backends more or less ready to go, but I wanted to ask what you think is the best way to get those changes merged into your branch. I could continue working on [this branch](https://github.com/jatinchowdhury18/RTNeural/tree/purefunctor-pure/implement-groups), but it's probably a bit out of date since your latest changes on this branch.

By the way, I think I'm also going to re-write the unit tests in Catch2 or something this weekend, so hopefully we can get this PR merged-in before then :).

**Comment by jatinchowdhury18** [comment]
Sorry, just getting back to this today. On [my branch](https://github.com/jatinchowdhury18/RTNeural/tree/purefunctor-pure/implement-groups) I've added:
- A TensorFlow test case
- STL non-templated implementation
- XSIMD implementations (template and non-templated)

Hopefully I can finish up the Eigen implementations tonight or tomorrow, and then we should be good to merge! I'd like to get the MicroTCN test case working for all the backends as well, but I'm okay with doing that after we merge this PR depending on when I get around to it.

**Comment by jatinchowdhury18** [comment]
Update: The Eigen implementation is now working as well, so the only work left to do should be some documentation and to finish up the MicroTCN test. That said, I think I'm going to wait until #119 is merged before merging these changes, just to avoid some tricky merge conflicts.

**Comment by purefunctor** [comment]
I can merge your branch into this one so it's easier to do the merge into main 👍🏼

**Comment by jatinchowdhury18** [comment]
Alrighty! The [other branch](https://github.com/jatinchowdhury18/RTNeural/tree/purefunctor-pure/implement-groups) should be ready to merge, if you'd like to pull those changes into this PR. Also feel free to add yourself to the [contributors list](https://github.com/jatinchowdhury18/RTNeural#contributors) if you'd like!

**Comment by jatinchowdhury18** [comment]
Looks like everything is good to go except for formatting. I'll go ahead and merge, and I can take care of formatting later.

Thanks again for your work on this!
